src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,318 @@
+
+        Self::check_depth(index_value.depth())?;
+        self.update_depth(index_value.depth());
+
+        // add provided node and its sibling to the leaves set
+        self.leaves.insert(index_value);
+        let sibling_node_index = index_value.sibling();
+        self.leaves.insert(sibling_node_index);
+
+        // add provided node and its sibling to the nodes map
+        self.nodes.insert(index_value, value);
+        self.nodes.insert(sibling_node_index, path[0].into());
+
+        // traverse to the root, updating the nodes
+        let mut index_value = index_value;
+        let node = Rpo256::merge(&index_value.build_node(value, path[0].into()));
+        let root = path.iter().skip(1).copied().fold(node, |node, hash| {
+            index_value.move_up();
+            // insert calculated node to the nodes map
+            self.nodes.insert(index_value, node);
+
+            // if the calculated node was a leaf, remove it from leaves set.
+            if self.leaves.contains(&index_value) {
+                self.leaves.remove(&index_value);
+            }
+
+            let sibling_node = index_value.sibling();
+            // node became a leaf only if it is a new node (it wasn't in nodes map)
+            if !self.nodes.contains_key(&sibling_node) {
+                self.leaves.insert(sibling_node);
+            }
+
+            // insert node from Merkle path to the nodes map
+            self.nodes.insert(sibling_node, hash.into());
+
+            Rpo256::merge(&index_value.build_node(node, hash.into()))
+        });
+
+        // if the path set is empty (the root is all ZEROs), set the root to the root of the added
+        // path; otherwise, the root of the added path must be identical to the current root
+        if self.root() == EMPTY_DIGEST {
+            self.nodes.insert(ROOT_INDEX, root);
+        } else if self.root() != root {
+            return Err(MerkleError::ConflictingRoots([*self.root(), *root].to_vec()));
+        }
+
+        // self.update_leaves()?;
+
+        Ok(())
+    }
+
+    /// Updates value of the leaf at the specified index returning the old leaf value.
+    ///
+    /// This also recomputes all hashes between the leaf and the root, updating the root itself.
+    pub fn update_leaf(
+        &mut self,
+        node_index: NodeIndex,
+        value: RpoDigest,
+    ) -> Result<RpoDigest, MerkleError> {
+        // check correctness of the depth and update it
+        Self::check_depth(node_index.depth())?;
+        self.update_depth(node_index.depth());
+
+        // insert NodeIndex to the leaves Set
+        self.leaves.insert(node_index);
+
+        // add node value to the nodes Map
+        let old_value = self.nodes.insert(node_index, value).unwrap_or(EMPTY_DIGEST);
+
+        // if the old value and new value are the same, there is nothing to update
+        if value == old_value {
+            return Ok(value);
+        }
```

Comment:
It is not clear to me why we are defaulting to `EMPTY_DIGEST` here. If we are trying to insert an `EMPTY_DIGEST` as a leaf, wouldn't this cause incorrect behavior?

Reply:
Yes, my bad. I forgot to update this code from the moment when I mistakenly thought that all empty nodes were taken as zero. In that case, should it return `Result<Option<RpoDigest>, MerkleError>` instead of `Result<RpoDigest, MerkleError>`, since old value can be `None`?